doc/simulation_environment.md
# Simulation Environment
This environment supports both parallel and perpendicular parking, using either discrete or continuous action spaces.  
As illustrated in the figure below, the custom environment can render a 2D environment with a top-down view and simulate parking movement using front-wheel steering through the Kinematic bicycle model. 
At each step, the car is rendered based on the input actions, which include acceleration and steering angle, and the next state of the vehicle is simulated.  

## Environment Description
The yellow rectangles represent the parking lot for two obstacles depicted as grey rectangles. 
The red rectangle indicates the parking lot for the agent, shown as the green rectangle.
The grey grid line is drawn at 1-meter intervals.
Although the visualization window size is 800 by 600 pixels, this corresponds to a physical size of 40 meters by 30 meters, defining 1 pixel as 0.05 meters and units are unified as meters. 
The parking lot size is 6 meters in length and 4 meters in width, while the car size for the agent is 4 meters in length and 2 meters in width. 
Furthermore, it can display some information such as the center of the car, the velocity, the heading angle and its radian. 
This is useful for debugging to make sure the car’s movement.

・Perpendicular parking image  
![image](https://github.com/taka-rl/RL_APS/assets/157423802/ae308d2d-1c6a-4a6c-a707-f5dae83b23db)

・Parallel parking image  
![image](https://github.com/taka-rl/RL_APS/assets/157423802/34acb574-9777-4ce0-91b5-6789214f6dda)

## Kinematic Bicycle Model
- Kinematic bicycle model equation  
  In order to simulate the car’s movement, the kinematic bicycle model was used in the environment. (x, y) are the coordinates of the center of the car. The car’s velocity is controlled within 10 km/m since the car is usually at a low speed during parking.
  This means if the velocity becomes over 10 km/m, it is clipped as 10 km/m. The following equation is the Kinematic bicycle model used in the simulation.
   
        x_dot = v * cos(ψ)
        y_dot = v * sin(ψ)
        v_dot = a
        ψ_dot = v * tan(delta) / CAR_L
  v: velocity, ψ(psi): heading angle of the car, a: acceleration, δ(delta): steering angle, CAR_L: car length  
  x: the center of the car position in x axle, y: the center of the car position in y axle
  
  The steering angle is limited between −𝜋/4 and 𝜋/4 to make sure the car's actions are realistic and to prevent the agent's exploration range from being too large.
  Therefore, in a given time, according to the equations above, the coordinate of the center of the car (𝑥′, 𝑦′) after moving can be obtained by the following equations.

      x_dash = dt ∗ x_dot + x  
      y_dash = dt ∗ y_dot + y  
      v = v_dot + v  
      ψ = dt ∗ ψ_dot  
  
## Reinforcement Learning
### Algorithm
Proximal Policy Optimization (PPO) was used for training the reinforcement learning agent. Experiment tests for PPO conducted by OpenAI prove that PPO outperforms other online policy gradient methods while keeping a favorable balance between sample complexity, simplicity, and wall-time. PPO is an actor-critic algorithm which takes a hybrid approach combining a value-based and policy-based approach, where the actor decides actions to take, and the critic evaluates the actions taken by the actor.   

### Environment
The overview of RL model for this project is illustrated in the following figure.  
![image](https://github.com/taka-rl/RL_APS/assets/157423802/192dc6cb-3ee7-4fec-8e92-db4cdd4a516c)

### Action type
There are two action types which are continuous and discrete.  
Actions as input values for the agent are [a, δ] in both type.  
a is acceleration, limited between -1 and 1 m/s^2  as the maximum value.
δ is set to between -𝜋/4〜𝜋/4 as the maximum value.
- continuous
  The agent can choose between -1 and 1 for both acceleration and steering angle, multiplying them by the maximum limit values.  

- discrete
  There are 6 different actions. 

  | number | action values[𝑎, δ] | description         |
  |--------|----------------------|---------------------|
  | 0      | [1, 0]               | move forward        |
  | 1      | [1, -𝜋/6]           | move right forward  |
  | 2      | [1, 𝜋/6]            | move left forward   |
  | 3      | [-1, 0]              | move backward       |
  | 4      | [-1, -𝜋/6]          | move right backward |
  | 5      | [-1, 𝜋/6]           | move left backward  |
 
### State value for the agent
This section describes the state values designed in this project.  

| State type                                        | Description                                                                                                                                                                                                                                |
|---------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| The coordinate of the parking space corner points | The transformed coordinate system from the global coordinate system to the local coordinate system and divided by the maximum distance defined as 'max_distance' in Config class in `parameters.py` for normalization.                     |
| Guidance point                                    | The center point of the parking lot in the coordinate system of the car, transformed into the car's relative coordinate system and by the maximum distance defined as 'max_distance' in Config class in `parameters.py` for normalization. | 
| Velocity                                          | The car's velocity value and is divided by the velocity limit defined as 'velocity_limit' in Config class in `parameters.py` for normalization.                                                                                            |

#### Description about relative coordinate system
In global coordinates, the car must account for its own position and orientation within the global frame, complicating calculations. 
Expressing a global coordinate system as a local coordinate system simplifies the representation, making it easier to manage and understand. 
The following figure illustrates the local coordinate system where the front side of the car is positive along the y-axis, and the right side of the car is positive along the x-axis. 
The distance between each parking lot vertex and the center of the car is transformed into the car’s relative coordinate system.

![image](https://github.com/taka-rl/RL_APS/assets/157423802/dc39da5f-558b-4350-ae86-c97411c1dfbf)


### Reward type
The following reward functions are designed in this project.
The reward is given at the end of each episode. Therefore, the current episode is terminated when one of these events happens.  
There are 4 types as reward types. Common is applied to all the types. 


| Reward name       | Type    | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
|-------------------|---------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Successful reward | Common  | When the agent reverses into the parking lot, then the agent obtains a +1 reward.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| Collision check   | Common  | When the agent enters the grey rectangle, the agent obtains a -1 reward.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | 
| Maximum step      | Common  | When the agent takes more than the set maximum step, it obtains a -1 reward.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
| Line cross-check  | Common  | When the agent crosses the parking lot border vertically or horizontally, it obtains a -1 reward. This check is implemented to ensure realistic behavior, preventing the car from crossing to the opposite side of the parking lot to park. It is dependent on the placement of the parking lot. If the agent crosses the bottom border, it indicates a horizontal border crossing. Another example is if the parking lot is placed on the right side of the visualization window; when the agent crosses the right border of the parking lot, it indicates a vertical border crossing.                                  |
| Maximum distance  | Common  | When the agent is farther away from the parking lot than the set maximum distance, it obtains a -1 reward.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
| Guidance reward   | type2/4 | When the car’s vertices are within the parking lot and the center of the car is within a set distance threshold from the center of the parking lot, the agent receives a +1 reward. A small value is then subtracted from this reward based on the angle error to ensure the car is parallel with the parking lot borders. The penalty for angle error is linearly related, where an angle error of 0 degrees results in no penalty, and larger errors reduce the reward accordingly. The purpose of this reward is to encourage the agent to park near the center of the parking lot and to be parallel to the borders. |
| Velocity penalty  | type3/4 | This velocity penalty is linearly related. If velocity is 0km/h, it is no penalty. Larger velocity value reduce the reward accordingly. The purpose of this penalty is to encourage the agent to approach the parking lot gently.                                                                                                                                                                                                                                                                                                                                                                                        |



sim_env/parameters.py
from typing import Union, Tuple
import numpy as np

PI = np.pi
PIXEL_TO_METER_SCALE = np.float32(0.05)


class CarSize:
    """
    Represents the dimensions of the car.

    Attributes:
        length (np.float32): Length of the car in meters (default: 4.0m).
        width (np.float32): Width of the car in meters (default: 2.0m).
        car_struct (np.ndarray): 2D array defining the car's four corner coordinates in meters.
        car_struct_2 (np.ndarray): Alternative 2D array representation for perpendicular parking.

    Default car size:
        - Length: 4 meters
        - Width: 2 meters
    """
    def __init__(self, length: float = 4.0, width: float = 2.0):
        self.length = np.float32(length)
        self.width = np.float32(width)
        self.car_struct = np.array([[+self.length / 2, +self.width / 2],
                                    [+self.length / 2, -self.width / 2],
                                    [-self.length / 2, -self.width / 2],
                                    [-self.length / 2, +self.width / 2]],
                                   dtype=np.float32)

        self.car_struct_2 = np.array([[+self.width / 2, +self.length / 2],
                                      [+self.width / 2, -self.length / 2],
                                      [-self.width / 2, -self.length / 2],
                                      [-self.width / 2, +self.length / 2]],
                                     dtype=np.float32)


class WheelSize:
    """
    Represents the dimensions and positions of the car's wheels.

    Attributes:
        length (np.float32): Length of a single wheel in meters (default: 0.75m).
        width (np.float32): Width of a single wheel in meters (default: 0.35m).
        wheel_struct (np.ndarray): 2D array defining the wheel's four corner coordinates in meters.
        wheel_pos (np.ndarray): Coordinates of the four wheels relative to the car.

    Default wheel size:
        - Length: 0.75 meters
        - Width: 0.35 meters

    Wheel positions:
        - Top right: (1.25, 0.75)
        - Bottom right: (1.25, -0.75)
        - Bottom left: (-1.25, -0.75)
        - Top left: (-1.25, 0.75)
    """

    def __init__(self, length: float = 0.75, width: float = 0.35):
        self.length = np.float32(length)
        self.width = np.float32(width)

        self.wheel_struct = np.array([[+self.length / 2, +self.width / 2],
                                      [+self.length / 2, -self.width / 2],
                                      [-self.length / 2, -self.width / 2],
                                      [-self.length / 2, +self.width / 2]],
                                     dtype=np.float32)
        self.wheel_pos = np.array([[1.25, 0.75],
                                   [1.25, -0.75],
                                   [-1.25, -0.75],
                                   [-1.25, 0.75]],
                                  dtype=np.float32)


class ParkingLotSize:
    """
    Represents different parking lot configurations.

    Attributes:
        length (np.float32): Length of the parking lot in meters (default: 6.0m).
        width (np.float32): Width of the parking lot in meters (default: 4.0m).
        parallel_horizontal (np.ndarray): Array defining the parking structure for horizontal parallel parking.
        parallel_vertical (np.ndarray): Array defining the parking structure for vertical parallel parking.
        perpendicular_horizontal (np.ndarray): Array defining the parking structure for
                                                horizontal perpendicular parking.
        perpendicular_vertical (np.ndarray): Array defining the parking structure for vertical perpendicular parking.
        offset_parallel (np.float32): Offset distance for static obstacles in parallel parking.
        offset_perpendicular (np.float32): Offset distance for static obstacles in perpendicular parking.

    Default parking lot size:
        - Length: 6 meters
        - Width: 4 meters
    """

    def __init__(self, length: float = 6.0, width: float = 4.0):
        self.length = np.float32(length)
        self.width = np.float32(width)

        self.parallel_horizontal = np.array([
            [+self.length / 2, +self.width / 2],
            [+self.length / 2, -self.width / 2],
            [-self.length / 2, -self.width / 2],
            [-self.length / 2, +self.width / 2]],
            dtype=np.float32)

        self.parallel_vertical = np.array([
            [+self.width / 2, +self.length / 2],
            [+self.width / 2, -self.length / 2],
            [-self.width / 2, -self.length / 2],
            [-self.width / 2, +self.length / 2]],
            dtype=np.float32)

        self.perpendicular_horizontal = np.array([
            [+self.width / 2, +self.length / 2],
            [+self.width / 2, -self.length / 2],
            [-self.width / 2, -self.length / 2],
            [-self.width / 2, +self.length / 2]],
            dtype=np.float32)

        self.perpendicular_vertical = np.array([
            [+self.length / 2, +self.width / 2],
            [+self.length / 2, -self.width / 2],
            [-self.length / 2, -self.width / 2],
            [-self.length / 2, +self.width / 2]],
            dtype=np.float32)

        self.offset_parallel = np.float32(self.length)
        self.offset_perpendicular = np.float32(self.width)


class RenderConfig:
    """
    Stores configuration parameters for rendering.

    Attributes:
        fps (int): Frames per second for rendering.
        window_width (int): Width of the simulation window in pixels.
        window_height (int): Height of the simulation window in pixels.
        grid_size (int): Grid size for rendering in pixels.
        colors (dict): Dictionary containing RGB tuples for different UI elements.

    """
    def __init__(self, fps: int = 30, window_width: int = 800, window_height: int = 600, grid_size: int = 20) -> None:
        self.fps = fps
        self.window_width = window_width
        self.window_height = window_height
        self.grid_size = grid_size

        self.colors = {
            "RED": (255, 100, 100),
            "GREEN": (0, 255, 0),
            "BLUE": (100, 200, 255),
            "YELLOW": (200, 200, 0),
            "BLACK": (0, 0, 0),
            "GREY": (100, 100, 100),
            "WHITE": (255, 255, 255),
            "GRID_COLOR": (200, 200, 200)
        }


class Config:
    """
    Stores all configuration parameters for the parking environment.

    Attributes:
        car_size (CarSize): Car size settings.
        wheel_size (WheelSize): Wheel size settings.
        parking_lot_size (ParkingLotSize): Parking lot size settings.

        reward_type (str): Reward type used in reinforcement learning (default: 'type1').
            - type1: Default reward, primarily based on parking success.
            - type2: Includes a guidance reward, encouraging the agent to park parallel to the parking lot.
            - type3: Contains a velocity penalty, encouraging the agent to approach the parking lot gently.
            - type4: Comprise both the guidance reward and velocity penalty, making the agent park more realistic.

        penalty_ratio (dict): Penalty ratio for both guidance reward and velocity penalty.

        state_type (str): State representation type used in reinforcement learning (default: 'type1').
            - type1: The agent receives relative coordinate data for each (x, y) vertex between the parking lot and
                        the agent location. It has 8 elements.
            - type2: The agent receives relative coordinate data for each (x, y) vertex between the parking lot and
                        the agent location, plus an additional guidance distance to help with alignment.
                        It has 10 elements.
            - type3: The agent receives relative coordinate data for each (x, y) vertex between the parking lot and
                        the agent location, along with its velocity. It has 9 elements.
            - type4: The agent receives relative coordinate data for each (x, y) vertex between the parking lot and
                        the agent location, along with the guidance distance and its velocity. It has 11 elements.

        acceleration_limit (np.float32): Maximum acceleration limit in m/s².
        steering_limit (np.float32): Maximum steering angle limit in radians.
        velocity_limit (np.float32): Maximum velocity limit in m/s.
        dt (np.float32): Time step used in simulation.
        max_distance (np.float32): Maximum distance for parking in meters.
        max_steps (int): Maximum number of steps per episode.

        max_angle_error (np.float32): Maximum allowable angle error for guidance reward in radians.
        center_threshold (np.float32): Threshold for parking center alignment for guidance reward in meters.

        window_width_offset (int): Offset to keep parking within the screen boundaries in pixels.
        window_height_offset (int): Offset to keep parking within the screen boundaries in pixels.

        default_parking_locations (dict): Dictionary mapping parking lot sides (1-4) to their respective locations.
        side (Union[int, Tuple[int, ...]]): Defines which side of the environment the parking lot is placed.
                                            It can be a single integer(fixed side) or a tuple(randomized side selection)
                                                1: Bottom side.
                                                2: Top side.
                                                3: Left side.
                                                4: Right side.
        car_loc_randomize_range (tuple): A tuple specifying the range (in meters) for randomizing
                                            the car's initial position. The first and second elements represent
                                            the minimum and maximum values, respectively.
        initial_distance_range (tuple): A tuple specifying the range (in meters) for setting the initial distance
                                            between the car and the parking lot. The first and second elements
                                            represent the minimum and maximum values, respectively.
        heading_angle_range (dict): Dictionary defining possible initial heading angles for the car in radians.

    """

    def __init__(self,
                 car_length: float = 4.0, car_width: float = 2.0,
                 wheel_length: float = 0.75, wheel_width: float = 0.35,
                 parking_length: float = 6.0, parking_width: float = 4.0,
                 max_distance: float = 25.0, max_steps: int = 80,
                 acceleration_limit: float = 1.0, steering_limit: float = PI/4, velocity_limit: float = 2.77778,
                 max_angle_error: float = PI/12, center_threshold: float = 1.0,
                 penalty_ratio: dict = None, reward_type: str = 'type1', state_type: str = 'type1',
                 side: Union[int, Tuple[int, ...]] = 1, default_parking_locations: dict = None,
                 car_loc_randomize_range: tuple = (-5, 5), initial_distance_range: tuple = (7.5, 15.0),
                 heading_angle_range: dict = None
                 ):
        self.car_size = CarSize(car_length, car_width)
        self.wheel_size = WheelSize(wheel_length, wheel_width)
        self.parking_lot_size = ParkingLotSize(parking_length, parking_width)

        # Reward and State settings
        self.reward_type = reward_type
        self.state_type = state_type
        self.penalty_ratio = {'angle': 0.25, 'velocity': 0.25} if penalty_ratio is None else penalty_ratio

        # Action limits
        self.acceleration_limit = np.float32(acceleration_limit)
        self.steering_limit = np.float32(steering_limit)
        self.velocity_limit = np.float32(velocity_limit)
        self.dt = np.float32(0.1)
        self.max_distance = np.float32(max_distance)
        self.max_steps = max_steps

        # Guidance reward
        self.max_angle_error = np.float32(max_angle_error)
        self.center_threshold = np.float32(center_threshold)

        # Rendering settings
        self.render_config = RenderConfig()
        self.window_width_offset = 100
        self.window_height_offset = 50

        # Parking lot and Car location, and car's heading angle for training
        self.default_parking_locations = default_parking_locations or {1: np.array([15.0, 2.5]),
                                                                       2: np.array([15.0, 27.5]),
                                                                       3: np.array([2.5, 15.0]),
                                                                       4: np.array([37.5, 15.0])}

        if isinstance(side, int):
            if side not in (1, 2, 3, 4):
                raise ValueError("side must be an integer between 1 and 4.")
        elif isinstance(side, tuple):
            if not all(isinstance(s, int) and s in (1, 2, 3, 4) for s in side):
                raise ValueError("Each value in side tuple must be an integer between 1 and 4.")
        else:
            raise TypeError("side must be either an int (fixed side) or a tuple (randomized side selection).")
        self.side = side
        self.car_loc_randomize_range = car_loc_randomize_range
        self.initial_distance_range = initial_distance_range
        self.heading_angle_range = heading_angle_range or {
                    "perpendicular": {
                        1: (PI/12 * 5, PI/12 * 7),
                        2: (-PI/12 * 7, -PI/12 * 5),
                        3: (-PI/12, PI/12),
                        4: (PI - PI/12, PI + PI/12),
                    },
                    "parallel": {
                        1: (PI/6, PI/3),
                        2: (-PI/3, -PI/6),
                        3: (-PI/12, PI/12),
                        4: (-PI/12 * 11, PI/12 * 11),
                    }
                }
